$OpenBSD: patch-kitty_desktop_c,v 1.1.1.1 2020/09/07 12:19:52 thfr Exp $

revert the use of dlopen(3) to load libcanberra
see commit 8184ba246a63b05cd79e742c48f5527ac29c8a67 (23-Oct-2019)

Index: kitty/desktop.c
--- kitty/desktop.c.orig
+++ kitty/desktop.c
@@ -7,6 +7,7 @@
 
 #include "data-types.h"
 #include <dlfcn.h>
+#include <canberra.h>
 
 #define FUNC(name, restype, ...) typedef restype (*name##_func)(__VA_ARGS__); static name##_func name = NULL
 #define LOAD_FUNC(handle, name) {\
@@ -97,64 +98,15 @@ static PyMethodDef module_methods[] = {
     {NULL, NULL, 0, NULL}        /* Sentinel */
 };
 
-static void* libcanberra_handle = NULL;
-static void *canberra_ctx = NULL;
-FUNC(ca_context_create, int, void**);
-FUNC(ca_context_destroy, int, void*);
-typedef int (*ca_context_play_func)(void*, uint32_t, ...); static ca_context_play_func ca_context_play = NULL;
+static ca_context *canberra_ctx = NULL;
 
-static PyObject*
-load_libcanberra_functions(void) {
-    LOAD_FUNC(libcanberra_handle, ca_context_create);
-    LOAD_FUNC(libcanberra_handle, ca_context_play);
-    LOAD_FUNC(libcanberra_handle, ca_context_destroy);
-    return NULL;
-}
-
-static void
-load_libcanberra(void) {
-    static bool done = false;
-    if (done) return;
-    done = true;
-    const char* libnames[] = {
-#if defined(_KITTY_CANBERRA_LIBRARY)
-        _KITTY_CANBERRA_LIBRARY,
-#else
-        "libcanberra.so",
-        // some installs are missing the .so symlink, so try the full name
-        "libcanberra.so.0",
-        "libcanberra.so.0.2.5",
-#endif
-        NULL
-    };
-    for (int i = 0; libnames[i]; i++) {
-        libcanberra_handle = dlopen(libnames[i], RTLD_LAZY);
-        if (libcanberra_handle) break;
-    }
-    if (libcanberra_handle == NULL) {
-        fprintf(stderr, "Failed to load %s, cannot play beep sound, with error: %s\n", libnames[0], dlerror());
-        return;
-    }
-    load_libcanberra_functions();
-    if (PyErr_Occurred()) {
-        PyErr_Print();
-        dlclose(libcanberra_handle); libcanberra_handle = NULL;
-    }
-    if (ca_context_create(&canberra_ctx) != 0) {
-        fprintf(stderr, "Failed to create libcanberra context, cannot play beep sound\n");
-        ca_context_destroy(canberra_ctx); canberra_ctx = NULL;
-        dlclose(libcanberra_handle); libcanberra_handle = NULL;
-    }
-}
-
 void
 play_canberra_sound(const char *which_sound, const char *event_id) {
-    load_libcanberra();
-    if (libcanberra_handle == NULL || canberra_ctx == NULL) return;
+    if (canberra_ctx == NULL) ca_context_create(&canberra_ctx);
     ca_context_play(
         canberra_ctx, 0,
-        "event.id", which_sound,
-        "event.description", event_id,
+        CA_PROP_EVENT_ID, which_sound,
+        CA_PROP_EVENT_DESCRIPTION, event_id,
         NULL
     );
 }
@@ -165,7 +117,6 @@ finalize(void) {
     libsn_handle = NULL;
     if (canberra_ctx) ca_context_destroy(canberra_ctx);
     canberra_ctx = NULL;
-    if (libcanberra_handle) dlclose(libcanberra_handle);
 }
 
 bool
